### ISSUE TITLE
[Pytorch] AMD GPUs benchmarks

### ISSUE THREAD

**Issue by hyperfraise** [ROCmSoftwarePlatform/pytorch#328]
Hi. Thanks for this work guys.

I was curious as to whether you had been able to bench the framework on amd gpus ? I've successfully build pytorch with rocm support following your instructions, and the benchs I got don't seem right. I'm testing with a Radeon 580, which should be like half the performance as 1080 Ti, and I'm seeing more like 9-10 times drop in performances on convolution. The tensorflow benchs already show that the gap shouldn't be that wide.

Is this supposed to be normal for the moment ?

**Comment by hyperfraise** [comment]
This is tensorflow benchmarks

http://blog.gpueater.com/en/2018/04/23/00011_tech_cifar10_bench_on_tf13/

**Comment by iotamudelta** [comment]
Thanks for the report! We are working on performance optimizations at the moment. If you are interested to help, could you `export MIOPEN_FIND_ENFORCE=3` prior to running the benchmark? This will tune the MIOpen convolution kernels for the benchmark (it will take a long time though!) - that way we could see what the "ideal" performance currently is.

**Comment by FelixSchwarz** [comment]
> it will take a long time though!

How long is "long" on a RX 580? Some hours? A day? A week?

**Comment by iotamudelta** [comment]
Typically a few hours for a single network. It depends on how many unique convolution configs are missing from our internal database.

**Comment by hyperfraise** [comment]
Ran this command with this bench : 
https://github.com/ryujaehun/pytorch-gpu-benchmark

Whole lotta tests were running. Waited the night for it to finish. Came back this morning and the task was frozen, radeontop showing every metric at 100% utilization (doubt it though). 

Looked for any result but none is showing. I'll try another, maybe lighter benchmark.

**Comment by hyperfraise** [comment]
Not sure if relevant for you, but I can't detect my gpu now, and this is the log of when I killed it when it froze during the first benchmark. Working on it.[
![Uploading 2018-12-12-112120_1127x1019_scrot.png…]()
](url)

**Comment by iotamudelta** [comment]
Couple of questions:
a) can you salvage the content of ~/.config/miopen and attach here? That'll at least give us an information how far it got before erroring out and if you keep the content of the file, the next time round no tuning for these configurations is required
b) which host kernel are you running this on and which dkms (or are you using upstream)?

Thanks!

**Comment by skylt** [comment]
(I'm working with @hyperfraise )
- a) [here](https://gist.github.com/skylt/2ff638c7f57b03d81178b3b028296b12)
- b) `4.15.0-42-generic` on ubuntu 18.04, with rocm `2.0.white-rabbit`

**Comment by iotamudelta** [comment]
@skylt Thanks, this is helpful. There is no `2.0.white-rabbit`. Could you provide `apt list | grep rocm`? Thanks!

The instability looks like a kernel driver issue on `gfx803`

**Comment by Delaunay** [comment]
I noticed that sometimes hcc is launched prior to the training beginning.
What is it doing ? Testing different kernels ? Compiling missing kernels ? 
Compiling the best kernels according to MIOpen's DB? 

The compilation step seems to disappear after the first script run. 
Is it correct to assume the timings after that first run should be consistent ?

You can find the results of the test I ran prior to tuning MIOpen below.
I am running the tuning at the moment (1 done, 3 to go).

The numbers are based on 90 observations (batch) where the first 10 were discarded.

GPU | Compute Speed (img/s) | Min | Max | f16 | batch | model
-- | -- | -- | -- | -- | -- | --
RX580 8GB | 379.5 | 264.53 | 670.77 | FALSE | 32 | resnet18
RX580 8GB | 345.8 | 300.10 | 405.88 | TRUE | 64 | resnet18
RX580 8GB | 153.4 | 121.76 | 208.56 | FALSE | 32 | resnet50
RX580 8GB | 150.1 | 137.98 | 167.72 | TRUE | 64 | resnet50
1060 6GB | 743.2 | 687.32 | 796.18 | FALSE | 32 | resnet18
1060 6GB | 975.6 | 845.16 | 1039.72 | TRUE | 64 | resnet18
1060 6GB | 155.1 | 153.46 | 156.16 | FALSE | 32 | resnet50
1060 6GB | 166.7 | 165.14 | 168.28 | TRUE | 64 | resnet50


GPU | Memory Used (Mo) | Memory Used (%) | Total Memory | f16 | batch | Model
-- | -- | -- | -- | -- | -- | --
RX580 8GB | 1793.08 | 21.90% | 8187 | FALSE | 32 | resnet18
RX580 8GB | 1834.20 | 22.40% | 8187 | TRUE | 64 | resnet18
RX580 8GB | 4196.18 | 51.26% | 8187 | FALSE | 32 | resnet50
RX580 8GB | 4321.52 | 52.79% | 8187 | TRUE | 64 | resnet50
1060 6GB | 2012.07 | 33.12% | 6075 | FALSE | 32 | resnet18
1060 6GB | 1999.51 | 32.91% | 6075 | TRUE | 64 | resnet18
1060 6GB | 4279.30 | 70.44% | 6075 | FALSE | 32 | resnet50
1060 6GB | 4272.04 | 70.32% | 6075 | TRUE | 64 | resnet50

**Comment by iotamudelta** [comment]
Concerning your questions: we do compile some kernels the first time we run them and subsequently get them from cache. So you are correct that the second time you'll invoke these kernels will be what you want to time.

I think the performance you observe makes sense. I'd expect resnet18 to get better as you tune MIOpen and in general I've observed that larger batch sizes help - so judging from your memory consumption numbers you could increase the batch size for resnet18 on the 580 2x (maybe even 4x). Please do attach your performance database here once you are done so that I can make sure we'll get these configs tuned in a future MIOpen release.

**Comment by geekboood** [comment]
> RX580 8GB	153.4	121.76	208.56	FALSE	32	resnet50

I wonder about how you get such a nice result. 
I use one RX480 but I only got about 83.92 img/sec.

**Comment by Delaunay** [comment]
Single GPU. I don't know where the differences could come from; I did not do anything in particular.
How did you measure the compute time ?

**Comment by geekboood** [comment]
I used the benchmark script in `https://github.com/ROCmSoftwarePlatform/pytorch/wiki/Performance-analysis-of-PyTorch`

**Comment by Delaunay** [comment]
I see; I did not run the same script and mine had a bug. You can find the updated numbers below.

I have not started to benchmark anything else yet. I will probably next week.
I would not expect everything to work perfectly yet. NVIDIA had a lot of time to tunes their library for a lot of different application that does not mean the underlying hardware is not capable of the same performance. 

I do think the convnets numbers show that their devices are capable. It is now a matter of optimization on the software level. For example the memory usage under ROCm is significantly lower than NVIDIA which suggests the algorithm under the hood are different and a trade off could be made between memory and speed.

The only big difference is f16 which I would think they did not yet work on specifically since none of their device really supports fast f16 operations (I think Radeon VII does but that's it). 

![img_sec](https://user-images.githubusercontent.com/4311937/53689998-d3539580-3d27-11e9-9287-531909aeb44d.png)
![memory_usage](https://user-images.githubusercontent.com/4311937/53689999-d3539580-3d27-11e9-818f-1d5d6527db50.png)

I also attached the MIOpen perf db.

[gfx803_36.cd.updb.txt](https://github.com/ROCmSoftwarePlatform/pytorch/files/2922686/gfx803_36.cd.updb.txt)

**Comment by iotamudelta** [comment]
@Delaunay this is very interesting data, thanks for posting it here! Which PyTorch commit did you use for your test? Your fp16 data looks correct for gfx803.

@geekboood could you post a link to the benchmark you were running? I'd be interested in having a look at it. Thanks!

**Comment by Delaunay** [comment]
The commit I used was `ffcbf1bd1` from `Fri Feb 22 16:36:06 2019 -0800`.
Yes this is for gfx803.

**Comment by iotamudelta** [comment]
Thanks, very good! It may be interesting to rerun this benchmark after `3ed44b67` is included for ROCm.

**Comment by geekboood** [comment]
@iotamudelta The ResNet one is already in the link above. 
As for the DrQA, I use this `https://github.com/facebookresearch/DrQA` code and train on CMRC dataset. You can also train the reader on Squad dataset. 
If you want to inspect the performance of ROCm, you can have a look at this one `https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/issues/173#issuecomment-469017319`. I have also done some trivial test on tensorflow-rocm.
